test/integration/zones_test.lua

```diff
@@ -0,0 +1,214 @@
+        }},
+    })
+    g.cluster:start()
+
+
+    g.A1 = g.cluster:server('A-1')
+    g.A2 = g.cluster:server('A-2')
+    g.A3 = g.cluster:server('A-3')
+end)
+
+g.after_all(function()
+    g.cluster:stop()
+    fio.rmtree(g.cluster.datadir)
+end)
+
+local function set_zones(zones)
+    local servers = {}
+    for k, v in pairs(zones) do
+        table.insert(servers, {uuid = k, zone = v})
+    end
+    return g.cluster.main_server.net_box:eval([[
+        local servers = ...
+        local ret, err = require('cartridge').admin_edit_topology({
+            servers = servers,
+        })
+        if ret == nil then
+            return nil, err
+        end
+        return true
+    ]], {servers})
+end
+
+local function set_distances(distances)
+    return g.cluster.main_server.net_box:call(
+        'package.loaded.cartridge.config_patch_clusterwide',
+        {{zone_distances = distances}}
+    )
+end
+
+local function get_config(...)
+    return g.cluster.main_server.net_box:call(
+        'package.loaded.cartridge.config_get_readonly', {...}
+    )
+end
+
+function g.test_zones()
+    local ok, err = set_zones({
+        [g.A1.instance_uuid] = 'z1',
+        [g.A2.instance_uuid] = 'z2',
+    })
+    t.assert_equals({ok, err}, {true, nil})
+
+    t.assert_items_equals(
+        g.A1:graphql({query = '{servers {uuid zone}}'}).data.servers,
+        {
+            {uuid = g.A1.instance_uuid, zone = 'z1'},
+            {uuid = g.A2.instance_uuid, zone = 'z2'},
+            {uuid = g.A3.instance_uuid, zone = box.NULL},
+        }
+    )
+
+    local ok, err = set_zones({
+        [g.A1.instance_uuid] = box.NULL, -- null doesn't edit value
+        [g.A2.instance_uuid] = '',
+    })
+    t.assert_equals({ok, err}, {true, nil})
+
+    local topology_cfg = get_config('topology')
+    t.assert_equals(topology_cfg.servers[g.A1.instance_uuid].zone, 'z1')
+    t.assert_equals(topology_cfg.servers[g.A2.instance_uuid].zone, nil)
+
+    local resp = g.A1:graphql({query = [[
+        mutation($servers: [EditServerInput]) {
+            cluster {
+                edit_topology(servers: $servers){
+                    servers {uuid zone}
+                }
+            }
+        }
+    ]], variables = {
+        servers = {
+            {uuid = g.A1.instance_uuid, zone = ''},
+            {uuid = g.A2.instance_uuid, zone = 'z2'},
+        }
+    }})
+
+    t.assert_equals(
+        resp.data.cluster.edit_topology.servers,
+        {
+            {uuid = g.A1.instance_uuid, zone = box.NULL},
+            {uuid = g.A2.instance_uuid, zone = 'z2'},
+        }
+    )
+
+    t.assert_error_msg_contains (
+        "bad argument params.zone to __edit_server"..
+        " (?string expected, got number)",
+        set_zones, {[g.A2.instance_uuid] = 2}
+    )
+end
+
+function g.test_distances()
+    local q_get_priority = [[
+        local replicaset = require('vshard').router.routeall()[...]
+        require('log').info(replicaset.priority_list)
```

Comment:
Remove if it's a debug. Add more verbose message if it's really needed.

Reply:
Fixed